### ISSUE TITLE
[Node.js] Express Server

### ISSUE THREAD

**Issue by shlee1353** [shlee1353/web-development-blog#12]
**Express** is a minimal and flexible Node.js web application framework that provides a robust set of features to develop web and mobile applications. It facilitates the rapid development of Node based Web applications. 

**tutorial source**
After cloning this repository, checkout to **feature/express-basic**

### Installing Express
```
npm init // package.json name : express-tutorial

npm install express --save

// npm install express-generator -g
// express myproject --ejs
```
### 1. 간단한 웹서버 작성
```js
// basic.server.js
var express = require('express');
var app = express();

app.use(function(request, response){
	response.send('<h1>hellp express</h1>');
});

app.listen(8080, function(){
	console.log('web server started');
});
```

### 2. 라우팅
- HTTP 요청의 메서드와 URL(PATH) 형태에 따라 분기하는 역할
- HTTP 메서드로부터 파생
```js
// route.server.js
var express = require('express');
var app = express();

app.get('/a', function(request, response){
	response.send('<a href="/b">Go to B</a>');
});

app.get('/b', function(request, response){
	response.send('<a href="/a">Go to A</a>');
});

// http://localhost:8080/page?name=Ryan
app.get('/page', function(request, response){
	response.send('welcome' + request.query.name);
});

// http://localhost:8080/page/Ryan
app.get('/page/:name', function(request, response){
	response.send('welcome' + request.params.name);
});

app.listen(8080, function(){
	console.log('web server started');
});
```

### 3. 미들웨어
- http 모듈과 express 모듈의 가장 큰 차이는 미들웨어의 유무
- 요청에 대한 응답을 완료하기 전까지 요청 중간에서 여러가지 일을 처리
- use() 메서드로 미들웨어를 적용
- next() 함수를 호출하여 다음 미들웨어로 제어권을 넘김
```js
// middleware.server.js
var express = require('express');
var app = express();

app.use('/special', function(request, response, next){
	console.log('스페셜 미들웨어');
	response.send('special middleware');
});

app.use(function(request, response, next){
	console.log('미들웨어 1');
	next();
});

app.use(function(request, response, next){
	console.log('미들웨어 2');
	response.end('middleware 1,2');
});

app.listen(8080, function(){
	console.log('web server started');
});
```
### 4. static 미들웨어
- express 에서 기본 제공하는 미들웨어
- 이미지, CSS, JS, HTML 파일과 같은 정적 리소스를 제공할 루트 경로를 지정하는 역할
```js
// static.server.js
var express = require('express');
var app = express();

// 기본호출
app.use(express.static(__dirname + '/static'));

//가상경로 지정
app.use('/public', express.static(__dirname + '/static'));

app.listen(8080, function(){
	console.log('web server started');
});
```
### 5. body-parser 미들웨어
- POST 요청시 전달된 데이터를 추출하는 미들웨어
- body-parser 미들웨어를 사용하려면 request객체에 body 속성이 부여됨
- [Postman](https://chrome.google.com/webstore/detail/postman/fhbjgbiflinjbdggehcddcbncdddomop)을 통해 POST 요청 확인 가능함
- 포스트맨 사용시 [Content-Type](https://gist.github.com/jays1204/703297eb0da1facdc454)은 application/x-www-form-urlencode를 사용 (key-value&key=value...의 형태로 전송)
```
npm install body-parser --save
```
```js
// static.server.js
...
var bodyParser = require('body-parser');

app.use(bodyParser.urlencoded({extended: false}));

app.post('/post', function(request, response){
	response.send("POST Data:" + request.body.content);
});
...
```
### 6. 템플릿 엔진
```
npm install ejs --save
```
```html
// index.ejs
<!DOCTYPE html>
<html>
<head>
	<title><%=title%></title>
</head>
<body>
	<h1><%=message%></h1>
</body>
</html>
```
```js
// template.js
var express = require('express');
var app = express();

app.set('views', './views');
app.set('view engine', 'ejs');
app.get('/', function(request, response){
	response.render('index', {title: 'Hello', message: 'Express!'});
});

app.listen(8080, function(){
	console.log('web server started');
});
```

### 7. todoapp API
After cloning this repository, checkout to **feature/express-todoapp**
**7-1. todo.server.js**
```js
// Express 모듈 추출
var express = require('express');

// body-parser 미들웨어 추출
var bodyParser = require('body-parser');

// Express 서버 객체 생성
var app = express();

// 정적 리소스 경로 지정
app.use(express.static(__dirname + '/static'));

// body-parser 미들웨어 적용
app.use(bodyParser.urlencoded({ extended: false }))

// 라우팅
app.use('/', require('./todo.router.js'));

// 서버 구동
app.listen(8080, function(){
	console.log('web server started');
});
```
To serve static files such as images, CSS files, and JavaScript files, use the **express.static** built-in middleware function in Express. You can load the files that are in the public directory. Express looks up the files relative to the static directory, so the name of the static directory is not part of the URL. **__dirname** is the directory name of the current module. It’s safer to use the absolute path of the directory that you want to serve:
```js
app.use(express.static('public'))

http://localhost:3000/images/kitten.jpg
http://localhost:3000/css/style.css
http://localhost:3000/js/app.js
http://localhost:3000/images/bg.png
http://localhost:3000/hello.html
```
To use multiple static assets directories, call the express.static middleware function multiple times:
```js
app.use(express.static('public'))
app.use(express.static('files'))
```

**7-2. todo.router.js**
```js
var fs = require('fs');
var express = require('express');
var router = express.Router();

// 데이터를 저장할 디렉토리
var dataPath = __dirname + '/data/';

// TODO 목록 조회
router.get('/todos', function(req, res){
	var files = fs.readdirSync(dataPath);
	// [ '1499652647873.json', '1499652721429.json', '1499652725947.json', '1499652729161.json', '1499652734901.json' ]
	var todos = [];
	// [ { id: 1499652647873, content: 'To do list 실습 테스트' },
	// { id: 1499652721429, content: '등록 기능 개발' },
	// { id: 1499652725947, content: '목록 조회 기능 개발' },
	// { id: 1499652729161, content: '삭제 기능 개발' },
	// { id: 1499652734901, content: '완성!' } ]

	files.forEach(function(file){
		var jsonData = JSON.parse(fs.readFileSync(dataPath + file, 'utf8'));
		todos.push(jsonData);
	});

	res.send({
		result: 'success',
		data: todos
	});
});

// TODO 추가
router.post('/todos', function(req, res){
	var id = new Date().getTime();
	// 1500880894981
	var filePath = dataPath + id + '.json';
	// .../data/1500880942949.json

	var jsonData = {
		id:id,
		content:req.body.content
	};
	// { id: 1500880894981, content: 'test' }

	fs.writeFileSync(filePath, JSON.stringify(jsonData), 'utf8');

	res.send({
		result: 'success',
		data: jsonData
	});
});

// TODO 삭제
router.delete(/todos\/(\d+)/, function(req, res){
	// \d is a digit (a character in the range 0-9), and + means 1 or more times. So, \d+ is 1 or more digits.
	try{
		fs.unlinkSync(dataPath + req.params[0] + '.json');
		res.send({
			result: 'success'
		});
	}catch(e){
		res.send({
			result: 'failed',
			error: e.syscall + ' error'
		});
	}
});


module.exports = router;
```
**express.Router**
모듈식 마운팅이 가능한 핸들러를 작성할 수 있음. 라우트를 모듈로 작성하고 미들웨어함수 로드 및 라우팅을 정의한 후 기본앱의 한 경로에 라우터 모듈을 마운트합니다.

The Node.js file system module allow you to work with the file system on your computer. To include the File System module, use the require() method:
- Read Files
```js
var fs = require('fs');
var express = require('express');
var app = express();

var dataPath = __dirname;

// readFile
fs.readFile(dataPath + '/index.html', 'utf8', function (err,data) {
  if (err) {
    return console.log(err);
  }
  console.log(data);
});

// readFileSync
var contensts = fs.readFileSync(dataPath + '/index.html').toString();
console.log(contensts);

app.listen(8080, function(){
	console.log('web server started');
});
```
- Create Files
```js
var fs = require('fs');
var express = require('express');
var app = express();

var dataPath = __dirname;
var filePath = dataPath + '/data/'

// Hello content!Hello content!Hello content!Hello content!
fs.appendFile(filePath + 'mynewfile1.txt', 'Hello content!', function (err) {
  if (err) throw err;
  console.log('Saved!');
});

// Hello content!
fs.writeFile(filePath + 'mynewfile2.txt', 'Hello content!', function (err) {
  if (err) throw err;
  console.log('Saved!');
});

app.listen(8080, function(){
	console.log('web server started');
});
```
- Delete Files
```js
var fs = require('fs');
var express = require('express');
var app = express();

var dataPath = __dirname;
var filePath = dataPath + '/data/'

fs.unlink(filePath + 'mynewfile2.txt', function (err) {
  if (err) throw err;
  console.log('File deleted!');
});
```
**7-3. app.js**
```js
var todoModel = (function($){
	function _api(url, method, data){
		var deferred = $.Deferred();
		
		$.ajax({
			url: url, 
			method: method, 
			data: data, 
			success: function(res){
				if(res.result === 'success'){
					deferred.resolve(res);
				}else{
					deferred.reject(res);
				}
			},
			error: function(err){
				console.warn('API 호출 실패 :', url, err);
			}
		});

		return deferred.promise();
	}

	return {
		getTodos : function(){
			return _api('/todos', 'get');
		}, // (9)

		addTodo : function(content){
			return _api('/todos', 'post', {content:content});
		},

		deleteTodo : function(id){
			return _api('/todos/' + id, 'delete');
		}
	};
})(jQuery);

var todoView = (function($){
	var self;
	var $els = {};
	
	var _prependTodo = function(id, content){
		var $newTodo = $els.todoItemTemplate.clone();

		$newTodo
			.attr('data-todoid', id)
			.find('._todoContent')
			.text(content);

		$els.todoList.prepend($newTodo);
	}; // (15)

	var _onSubmit = function(e){
		var content = $els.todoInput.val().trim();

		$els.todoInput
			.val('')
			.focus();

		if(content.length){
			$(self).trigger('addTodo', [content]);
		}

		e.preventDefault();
	}; // (12)

	var _onClickDeleteBtn = function(e){
		var todoId = $(e.currentTarget).closest('._todoItem').attr('data-todoid');

		if(confirm('삭제하시겠습니까?')){
			$(self).trigger('deleteTodo', [todoId]);
		}
	};

	return {
		init : function(){
			self = this;

			this.cacheElements();
			this.bindEvents();
		}, // (3)

		cacheElements : function(){
			$els.todoForm = $('._todoInputForm');
			$els.todoInput = $('._todoInput');
			$els.todoItemTemplate = $('._todoItem').detach();
			$els.todoList = $('._todoList');
			
			$els.todoList
				.empty()
				.show();
		}, // (4), (5)

		bindEvents : function(){
			$els.todoForm.on('submit', _onSubmit);
			$els.todoList.on('click', '._deleteBtn', _onClickDeleteBtn);
		}, // (6), (11)

		renderTodos : function(data){
			$els.todoList.empty();

			$(data).each(function(i, item){
				_prependTodo(item.id, item.content);
			});
		}, // (10)

		deleteTodo : function(id){
			$els.todoList.find('._todoItem[data-todoid='+id+']').remove();
		},

		prependTodo : function(id, content){
			_prependTodo(id, content);
		}
	}
})(jQuery);

var todoController = (function($, model, view){
	var self;

	var _onAddTodo = function(e, content){
		model
			.addTodo(content)
			.then(function(res){
				view.prependTodo(res.data.id, res.data.content);
			});					
	}; // (14)

	var _onDeleteTodo = function(e, id){
		model
			.deleteTodo(id)
			.then(function(res){
				view.deleteTodo(id);
			});	
	};

	return {
		init : function(){
			self = this;

			view.init();

			this.bindEvents();
			this.renderTodos();
		}, // (2)
               // 뷰에서 발생한 특정 이벤트에 대해 특정 모델값을 업데이트 한 후 다시 뷰로 전달하는 바인딩
		bindEvents : function(){
			var $view = $(view);

			$view
				.on('addTodo', _onAddTodo)
				.on('deleteTodo', _onDeleteTodo);
		}, // (7), (13)

		renderTodos : function(){
			model
				.getTodos()
				.then(function(res){
					view.renderTodos(res.data);
				});
		} // (8)

	};
})(jQuery, todoModel, todoView).init(); // (1)
```
**Initial Setting**
1. todoController.init() 실행
2. view.init(), bindEvents(), renderTodos() 실행
3. cacheElements(), bindEvents() 실행
4. The **.detach()** method is the same as .remove(), except that .detach() keeps all jQuery data associated with the removed elements. This method is useful when removed elements are to be reinserted into the DOM at a later time.
5. Object {todoForm: n.fn.init(1), todoInput: n.fn.init(1), todoItemTemplate: n.fn.init(1), todoList: n.fn.init(1)}
6. bindEvents() 실행 'submit', 'click' 이벤트 바인딩
7. jQuery 메소드를 사용하기 위해 view 객체를 $() 적용후 'addTodo', 'deleteTodo' 이벤트 바인딩
8. mode.getTodos() 실행후 res 값을 받아, view.renderTodos(res.data)로 전달. res는 Object {result: "success", data: Array(6)}
9. res값은 _api('/todos', 'get')를 통해 가져온 데이터 값임. 이때 비동기 프로그래밍을 위해 var deferred = $.Deferred() 를 사용함. $.Deferred()로 deferred 객체를 만들고, 성공했을 때에는 resolve, 실패했을 때에는 reject 메소드를 호출하면 resolve는 done으로, reject는 fail로 연결됨. done이나 fail로 구분하지 않고 한 번에 처리하려면 then 메소드 사용. longAndComplicatedFunction 함수에서 deferred.promise()를 return 해야함.
```js
var longAndComplicatedFunction = function() {
	var deferred = $.Deferred();
	try {
		// 완료되려면 50초가 걸리는 매우 복잡한 비동기 코드
		deferred.resolve('성공');
	} catch (err) {
    	       deferred.reject(err);
  	}
	return deferred.promise();
};
longAndComplicatedFunction().done(function(message) {
	console.log(message);
}).fail(function(error) {
	console.log(error);
}).always(function() {
	console.log('완료!');
});
console.log('다음 행동');
```
10. renderTodos() 실행. 기존에 있던 리스트($els.todoList.empty())를 삭제 후 전달받은 값을 이용해 each 구문실행. 'data-todoid' 값과 컨텐츠를 담고있는 아이템을 생성한후 리스트에 추가한다 ($els.todoList.prepend($newTodo))

**add item**
11. submit 이벤트 발생 후, _onSubmit 함수 실행
12. input의 값을 content에 담고, 초기화. content 값이 있을 경우 $(self).trigger('addTodo', [content]) 실행. addTodo 는 커스텀 이벤트로 .on 으로 연결된 함수를 실행
```js
$( "#foo" ).on( "custom", function( event, param1, param2 ) {
  alert( param1 + "\n" + param2 );
});
$( "#foo").trigger( "custom", [ "Custom", "Event" ] );
```
13. $view.on('addTodo', _onAddTodo) 실행. 
14. 입력한 내용은 [POST data](https://kevinchisholm.gitbooks.io/handling-post-requests-with-express-and-node-js/content/)를 통해 Node.js 웹서버로 전달되며, 전달된 값은 fs.writeFileSync()를 통해 시스템에 저장되며, 생성된 id와 content는 클라이언트 쪽으로 반환됩니다. 
15. view.prependTodo()의 인자로 전달 되어 _prependTodo()를 실행합니다.

참고
- https://zellwk.com/blog/crud-express-mongodb/
- https://mlab.com/
